### PR TITLE
Make Table use ListDataSource + Example Table app

### PR DESCRIPTION
**Pull request by Ocupe**
* Documentation and format of old in Google docstring format.
* modified cocoa `Table` to changes.

The table in the current version was not actually converting the input data into a `ListDataSource`. 
Now it does but there is a ugly and confusing naming problem. If we want to get the data from the table widget we have to do this:

~~~python
# get `ListDataSource` from the table
>>> table.data
# get the actual data from the table
>>> table.data.data
~~~

I don't know if we should rename `table.data` to `table.data_source` or do something else?
But then we would have to so this:
~~~python
# add data to the table
>>> table.data_source = ['1', '2', '3']
# get data from the table
>>> table.data_source.data
['1', '2', '3']
~~~

There are also some parts that I commented out. Are these parts not needed any more with the new `ListDataSource` approach? @freakboy3742

### CODE REVIEW

File: src/cocoa/toga_cocoa/widgets/table.py [hunk 1]
```diff
@@ -7,27 +7,27 @@ class TogaTable(NSTableView):
...
     @objc_method
     def numberOfRowsInTableView_(self, table) -> int:
-        return len(self.interface.data)
+        return len(self.interface.data.data) if self.interface.data else 0
 
     @objc_method
     def tableView_objectValueForTableColumn_row_(self, table, column, row: int):
         column_index = int(column.identifier)
-        return self.interface.data[row][column_index]
+        return self.interface.data.data[row].data[column_index]
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/widgets/table.py]: This indicates a problem with the interface - there's a few too many datas here :-)

This should be behind an API like `self.interface.data.item(row, column_index)`. The data attribute of a ListDataSource shouldn't be exposed like this.

File: src/cocoa/toga_cocoa/widgets/table.py [hunk 2]
```diff
@@ -60,14 +60,14 @@ def create(self):
...
-    def remove_row(self, node):
-        del self.node[node._impl]
+    # def insert_row(self, node):
+    #     node._impl = TogaNodeData.alloc().init()
+    #     node._impl.node = node
+    #
+    #     self.node[node._impl] = node
+    #
+    # def remove_row(self, node):
+    #     del self.node[node._impl]
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/widgets/table.py]: This can be deleted. Tables shouldn't have a native insert method; insertion is done to the data source.

File: src/core/toga/widgets/table.py [hunk 3]
```diff
@@ -3,16 +3,15 @@
...
+            source: The ``class:ListDataSource`` that the row belongs to.
+            data: A ``list`` where each element is a column of the row.
+            icon: A icon displayed in the row.
+        """
 
-        :param data: Information about the row
-        :type  data: ``dict``
-        '''
         self._impl = None
         self.source = source
```
Comment (freakboy3742) [inline src/core/toga/widgets/table.py]: This is something that can be cleaned up - you shouldn't need to specify `data` *and* `source` - you should just need data.

File: src/core/toga/widgets/table.py [hunk 4]
```diff
@@ -100,60 +87,61 @@ class Table(Widget):
...
-            If no style is provided then a new one will be created for the widget.
+            If no style is provided` then a new one will be created for the widget.
+        on_select(``callable``): A function to be invoked on selecting a row of the table.
         factory (:obj:`module`): A python module that is capable to return a
             implementation of this class with the same name. (optional & normally not needed)
+
+    Examples:
+        >>> headings = ['Head 1', 'Head 2', 'Head 3']
+        >>> data = [['item 1', 'item 2', 'item3'],
+        >>>         ['item 1', 'item 2', 'item3']]
```
Comment (freakboy3742) [inline src/core/toga/widgets/table.py]: Although both will work, this should be a list of tuples - the inner items have position specific type.

**Comment by Ocupe** [comment]
@freakboy3742 this is ready for a final review.

File: examples/table/.gitignore [hunk 5]
```diff
@@ -0,0 +1,58 @@
+# Byte-compiled / optimized / DLL files
```
Comment (freakboy3742) [inline examples/table/.gitignore]: Is this file required?

File: src/core/toga/widgets/table.py [hunk 6]
```diff
@@ -24,136 +19,129 @@ def __repr__(self):
...
             self._data = ListDataSource(data)
         else:
             self._data = data
 
         if data is not None:
-            self._data.interface = self
-
-        self._impl.refresh()
+            # self._data.interface = self
+            self._impl.refresh()
```
Comment (freakboy3742) [inline src/core/toga/widgets/table.py]: Won't this be required even on the "data is None" case? If I clear the data source, won't I need to redraw the (now empty) widget?

File: examples/table/LICENSE [hunk 7]
```diff
@@ -0,0 +1,28 @@
+
+Copyright (c) 2017, Jonas Schell
```
Comment (freakboy3742) [inline examples/table/LICENSE]: Another case of "is this file required?". Since it's being added to an existing repository, we don't need sublicenses (unless you particularly wish to explicitly retain copyright to this piece of code)

**Comment by Ocupe** [comment]
@freakboy3742 Regarding the updating of the table on data change. Is it the responsibility of the `ListDataSource` to update the table when I change something on the data source (insert row, clear table ...) or has the programmer call some refresh method after he/she changed something? At the moment this functionality is missing.
I don't know if there is the possibility to 'watch' the data_source in the cocoa backend and detect changes to trigger a automatic update of the table?

My idea:
We could give the possibility to provide a `update_function` that is called every time the data for the table changed (insert, delete, clear).
~~~python
class ListDataSource:
    def __init__(self, data, refresh_function=None):
        self._data = self.create_rows(data)
        self.refresh = refresh_function if refresh_function else None
~~~

**Comment by freakboy3742** [comment]
`ListDataSource` is a wrapper providing a list-like API for manipulating data, but it doesn't *actually* have to be backed by a literal list. As a thought experiment, you could create a "Fibonacci" data source, where the value in each row can be established, but you don't have to create a list and store every value in the sequence.

The other use case to keep in mind - Having 2 visualisations of a single source of data. Think Cricket's tree of tests (a tree not a list in this case, but the idea is the same). There's one canonical list of tests; but two views - one "show everything" view, and one "show only the failed items" view. You shouldn't need to maintain two independent representations of the data - you only need a single underlying collection of data, with two data sources (one of which is filtered).

I suppose there *might* be a use case for two independent visualisations on a *single* data source... but that's going to get messy on notifications.

So - the idea about having a callback is absolutely on the right track. However, it's also implicitly covered when a DataSource is registered with a widget. See [data.setter] on the tree widget; it sets the interface for the data source - which then gives you the object to call back on. 

That said, making this an explicit callback function (and just passing `self` or `self.refresh`) in as the argument would allow the callback to be used creatively - passing through some *other* method on the way. The widget is a default callback handler; but if you want to use the data source outside widgets, or in conjunction with something else (say - update a blinkenlight every time the data source changes), you'd be able to do so. Thoughts?

File: src/core/toga/widgets/table.py [hunk 8]
```diff
@@ -25,127 +20,160 @@ def __repr__(self):
...
-        Set the data source of the data
-
-        :param data: Data source
-        :type  data: ``list``, ``tuple``, or ``class``
-        '''
         if isinstance(data, (list, tuple)):
-            self._data = ListDataSource(data)
+            self._data = ListDataSource(data, refresh_function=self._impl.refresh)
         else:
             self._data = data
```
Comment (phildini) [inline src/core/toga/widgets/table.py]: Given that we claim a promise above about what we accept, it would be nice to throw an error here if `data` is not an iterable or a ListDataSource.

...

Thinking about this more, the only promise necessary is that `self._data` is some kind of iterable, right? I mean, we could (theoretically) pass in a generator as the data source, couldn't we?

Could we check for that here?

**Comment by phildini** [comment]
I'm super excited about this. One Nit to consider, but then I think this is _real close_ to shippable.

File: src/core/toga/widgets/table.py [hunk 9]
```diff
@@ -25,127 +20,168 @@ def __repr__(self):
...
-        node = TreeNode(source=self, data=data, icon=icon)
+    @property
+    def refresh_list(self):
+        return self._refresh_list
+
+    def add_to_refresh_list(self, widget):
+        self._refresh_list.append(widget)
+
+    def remove_from_refresh_list(self, widget):
+        self._refresh_list.remove(widget)
```
Comment (freakboy3742) [inline src/core/toga/widgets/table.py]: `add_to_refresh_list` and `remove_from_refresh_list` are a little cumbersome... how about just using the Gang Of Four name for the pattern - "Listener" (`add_listener`, `remove_listener`, `listeners`). This could then be a common named used in other APIs (e.g., the TreeDataSource); when there are multiple possible events, the "listener" interface just gains more endpoints.

File: src/core/toga/widgets/table.py [hunk 10]
```diff
@@ -25,127 +20,168 @@ def __repr__(self):
...
+    Args:
+        data (`list` of `tuple`): A list of tuples containing the data for every row.
+        refresh_function (`callable`): A function invoked on data change.
+    """
+
+    def __init__(self, data, on_refresh=None):
         self._data = self.create_rows(data)
-        self.interface = None
+        self._on_refresh = on_refresh if on_refresh else None
+        self._refresh_list = []
```
Comment (freakboy3742) [inline src/core/toga/widgets/table.py]: What's the distinction between the "on_refresh" handler, and the "refresh_list" handlers? Aren't they the same thing? Couldn't we either (a) remove this on_refresh argument, or (b) make this a "listeners" argument which provides an initial list of listeners?

File: src/core/toga/widgets/table.py [hunk 11]
```diff
@@ -159,3 +195,6 @@ def on_select(self, handler):
         """
         self._on_select = wrapped_handler(self, handler)
         self._impl.set_on_select(self._on_select)
+
+    def refresh(self):
+        self._impl.refresh()
```
Comment (freakboy3742) [inline src/core/toga/widgets/table.py]: Does the Interface level table need a refresh? Is there a reason the *user* would need to invoke `refresh()`? If not, should we just keep this on the implementation level?

File: src/core/toga/widgets/table.py [hunk 12]
```diff
@@ -25,127 +20,168 @@ def __repr__(self):
...
+    """ A Table Widget allows the display of data in the from of columns and rows.
 
     Args:
         headings (``list`` of ``str``): The list of headings for the table.
         id (str): An identifier for this widget.
+        data (``list`` of ``tuple``): The data to be displayed on the table.
         style (:class:`colosseum.CSSNode`): An optional style object.
-            If no style is provided then a new one will be created for the widget.
+            If no style is provided` then a new one will be created for the widget.
+        on_select(``callable``): A function to be invoked on selecting a row of the table.
```
Comment (freakboy3742) [inline src/core/toga/widgets/table.py]: missing space after on_select?

File: src/core/toga/widgets/table.py [hunk 13]
```diff
@@ -25,127 +20,168 @@ def __repr__(self):
...
-        :returns: The data source of the tree
-        :rtype: ``dict``
-        '''
-        return self._data
+        """ The data source of the widget. It accepts table data
+        in the form of ``list``, ``tuple``, or :obj:`ListDataSource`
+
+        Returns:
+            (list) Returns a ``list`` of lists. Where the outer lists represents the
+            rows and each inner list represents a column.
```
Comment (freakboy3742) [inline src/core/toga/widgets/table.py]: Is this correct? Won't it return a DataSource?

File: src/dummy/toga_dummy/widgets/table.py [hunk 14]
```diff
@@ -5,12 +5,6 @@ class Table(Widget):
     def create(self):
         pass
 
-    def insert_row(self, node):
-        pass
-
-    def remove_row(self, node):
-        pass
-
```
Comment (freakboy3742) [inline src/dummy/toga_dummy/widgets/table.py]: This deletion should be reflected in the other platform implementations.

**Comment by Ocupe** [comment]
@freakboy3742 @phildini So, I made some changes that hopefully address your comments. The refresh_function is now called `on_refresh` and entirely for the user to play around with. 

A added a `self._refresh_list` to the `ListDataSource` to allow multiple tables to use the same data source. The idea is that table widgets subscribe to the data source to get updated on data change.

**Comment by Ocupe** [comment]
@freakboy3742 @phildini I hope I addressed all of your comments.

File: src/core/toga/widgets/table.py [hunk 15]
```diff
@@ -25,127 +20,168 @@ def __repr__(self):
...
-        :param data: Data source
-        :type  data: ``list``, ``tuple``, or ``class``
-        '''
         if isinstance(data, (list, tuple)):
             self._data = ListDataSource(data)
-        else:
+        elif isinstance(data, ListDataSource):
             self._data = data
+        else:
+            raise UserWarning('The table widgets should get data in form of a list or ListDataSource.')
```
Comment (freakboy3742) [inline src/core/toga/widgets/table.py]: We definitely don't want to be this prescriptive. The old construction would have allowed any user-created class that quacked the same as a ListDataSource to be used as a data source.

File: src/core/toga/widgets/table.py (hunk 8 above)
Comment (freakboy3742) [reply src/core/toga/widgets/table.py]: I disagree that this is needed. One of the use-cases for this is to pass in an arbitrary object that adheres to the ListDataSource interface, but isn't actually backed by a ListDataSource. As long as the data source quacks the same, we don't care what object it is.

**Comment by Ocupe** [comment]
@freakboy3742 I removed the `on_refresh` API because as you mentioned it is more or less the same functionality that is already provided by the `listeners` API. My intention was that the `listeners` stuff is handling subscriptions behind the scenes. Because you need to know that the object that you pass needs a `self._impl.refresh()` method to work. With the `on_refresh` property you could just pass any function as a callback.`

**Comment by Ocupe** [comment]
Maybe the last commit is a good way to solve the previously mentioned problem. 
Now you can also add functions as a listener not just a widget with the the right structure.
